File: Solution/CSharp/20.ValidParentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Stack<char> stack = new Stack<char>();
        foreach (var c in s) {
            if (stack.Count == 0) {
                stack.Push(c);
            } else {
                if (!IsPair(stack.Peek(), c)) {
                    stack.Push(c);
                } else {
                    stack.Pop();
                }
            }
        }
        return stack.Count == 0;
    }

    public bool IsPair(char c1, char c2) {
        if ((c1 == '(' && c2 == ')') || (c1 == ')' && c2 == '(')) {
            return true;
        } else if ((c1 == '{' && c2 == '}') || (c1 == '}' && c2 == '{')) {
            return true;
        } else if ((c1 == '[' && c2 == ']') || (c1 == ']' && c2 == '[')) {
            return true;
        }
        return false;
    }
}

File: Solution/Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        pair = {
            ")": "(",
            "]": "[",
            "}": "{"
        }

        stack = []
        for i in s:
            if i in pair.values():
                stack.append(i)
            elif i in pair.keys():
                if len(stack) == 0 or pair[i] != stack.pop():
                    return False
            else:
                return False

        return len(stack) == 0


solution = Solution()
s = "()[]{}"
res = solution.isValid(s)
print(res)
